[patch] Fix use of "index=N" when no injected files are present

mcb30 opened the pull request:
Commit 7f42b27 ("[patch] Patch the WIM image to include additional files") introduced a hidden bug: if the "index=N" parameter is used to modify the boot index but no injected files are present, then nothing will ever update the boot image metadata within the stored patched WIM header.  (When injected files are present, the injection code will subsequently update both the boot index and the boot image metadata as part of the injection mechanism.)

This bug remained hidden for many years.  Partly because not many people use the "index=N" feature, and partly because under UEFI the presence of the "wimboot" binary itself within the virtual filesystem exposed by iPXE meant that there would always be at least one injected file, and so the faulty code path would not be taken.

Commit ac4190a ("[patch] Avoid patching wimboot binary itself into image") fixed the unintended injection of the "wimboot" binary itself, thereby revealing the underlying bug in the existing code.

Fix by explicitly updating the boot image metadata within the stored patched WIM header, before considering the existence of injected files.

Reported-by: Anatoli Alexandrov <tony_alexandrov@hotmail.com>
Tested-by: Anatoli Alexandrov <tony_alexandrov@hotmail.com>
Signed-off-by: Michael Brown <mbrown@fensystems.co.uk>

NiKiZe commented:
Just a sanitycheck, index can never be zero?
Scenario where Image default boot index is != 0 and the cmdline index=0 ?

mcb30 commented:
> Just a sanitycheck, index can never be zero? Scenario where Image default boot index is != 0 and the cmdline index=0 ?

Thanks for reviewing! From memory, the indexes in WIM images start from 1, which is why we use 0 to mean "no override".

mcb30 commented:
> > Just a sanitycheck, index can never be zero? Scenario where Image default boot index is != 0 and the cmdline index=0 ?
> 
> Thanks for reviewing! From memory, the indexes in WIM images start from 1, which is why we use 0 to mean "no override".

... but in the process of making sure that I had remembered this correctly, I found a bug :slightly_smiling_face: - now fixed in commit https://github.com/ipxe/wimboot/commit/5be96c5

I've rebased this branch on top of current master (including this fix and a GCC 12 build fix).

Now just waiting for @rushyn to confirm that it works!

rushyn commented:
@mcb30 I have tested the fixindex brunch, its booting index selected.

mcb30 commented:
> @mcb30 I have tested the fixindex brunch, its booting index selected.

Thanks for testing: will merge now.

mcb30 commented:
@rushyn All merged; thanks for all your work.  This will be included in the v2.7.5 release, which has been submitted for Secure Boot signing.

rushyn commented:
@mcb30 I was happy to help. thank you for fixing the issue.
